health 데이터를 활용하여 **인종 별** **'질병이력의 합'**과 **'주관적 건강 인식'**의 상관계수를 구한 뒤, 상관계수가 **가장 큰 인종**과 **가장 작은 인종**을 구하세요.

1) 문자형 변수를 숫자형 변수로 바꿔서 진행할 것.

** 질병 이력 변수: HeartDisease(1), Diabetic(12), Stroke(5), KidneyDisease(17), SkinCancer(18), Asthma(16)

** 주관적 건강 인식: GenHealth(14) (poor < fair < good < very good < excellent)



In [1]:
getwd()

[1] "C:/Users/alsdu/dacon"

In [2]:
setwd("C:/Users/alsdu/OneDrive/문서/카카오톡 받은 파일/20220424_데이터분석 스터디/20220424_데이터분석 스터디")

In [16]:
df<-read.csv("health.csv")

In [4]:
str(df)

'data.frame':	319795 obs. of  18 variables:
 $ HeartDisease    : Factor w/ 2 levels "No","Yes": 1 1 1 1 1 2 1 1 1 1 ...
 $ BMI             : num  16.6 20.3 26.6 24.2 23.7 ...
 $ Smoking         : Factor w/ 2 levels "No","Yes": 2 1 2 1 1 2 1 2 1 1 ...
 $ AlcoholDrinking : Factor w/ 2 levels "No","Yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ Stroke          : Factor w/ 2 levels "No","Yes": 1 2 1 1 1 1 1 1 1 1 ...
 $ PhysicalHealth  : num  3 0 20 0 28 6 15 5 0 0 ...
 $ MentalHealth    : num  30 0 30 0 0 0 0 0 0 0 ...
 $ DiffWalking     : Factor w/ 2 levels "No","Yes": 1 1 1 1 2 2 1 2 1 2 ...
 $ Sex             : Factor w/ 2 levels "Female","Male": 1 1 2 1 1 1 1 1 1 2 ...
 $ AgeCategory     : Factor w/ 13 levels "18-24","25-29",..: 8 13 10 12 5 12 11 13 13 10 ...
 $ Race            : Factor w/ 6 levels "American Indian/Alaskan Native",..: 6 6 6 6 6 3 6 6 6 6 ...
 $ Diabetic        : Factor w/ 4 levels "No","No, borderline diabetes",..: 3 1 3 1 1 1 1 3 2 1 ...
 $ PhysicalActivity: Factor w/ 2 levels "N

In [5]:
head(df)

HeartDisease BMI   Smoking AlcoholDrinking Stroke PhysicalHealth MentalHealth
1 No           16.60 Yes     No              No      3             30          
2 No           20.34 No      No              Yes     0              0          
3 No           26.58 Yes     No              No     20             30          
4 No           24.21 No      No              No      0              0          
5 No           23.71 No      No              No     28              0          
6 Yes          28.87 Yes     No              No      6              0          
  DiffWalking Sex    AgeCategory Race  Diabetic PhysicalActivity GenHealth
1 No          Female 55-59       White Yes      Yes              Very good
2 No          Female 80 or older White No       Yes              Very good
3 No          Male   65-69       White Yes      Yes              Fair     
4 No          Female 75-79       White No       No               Good     
5 Yes         Female 40-44       White No       Yes              Very good
6 Yes         Female 75-79       Black No       No               Fair     
  SleepTime Asthma KidneyDisease SkinCancer
1  5        Yes    No            Yes       
2  7        No     No            No        
3  8        Yes    No            No        
4  6        No     No            Yes       
5  8        No     No            No        
6 12        No     No            No

In [10]:
#클래스 확인
table(df$Race)


American Indian/Alaskan Native                          Asian 
                          5202                           8068 
                         Black                       Hispanic 
                         22939                          27446 
                         Other                          White 
                         10928                         245212 

In [11]:
# Diabetic(12)
table(df[12])


                     No No, borderline diabetes                     Yes 
                 269653                    6781                   40802 
 Yes (during pregnancy) 
                   2559 

- 12번 변수만 질병이력 중 클래스 4개로 확인

# 1. 범주 재조정
- Diabetic(12)변수 클래스 조정 (No, Yes)

In [17]:
# head( df[,12] %in% "Yes" , 5 ) 
df[,12]<-ifelse(df[,12] %in% "Yes" , 1,0 )
table( df[12] ) 


     0      1 
278993  40802 

# 2. Label Encoding
## 2.1. 12번 제외 나머지 변수 (순서X)

In [18]:
install.packages("superml")

package 'superml' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\alsdu\AppData\Local\Temp\RtmpqO2lzf\downloaded_packages


In [19]:
library(R6)
library(superml) # Lebel Encoder

Warning message:
"package 'superml' was built under R version 3.6.3"

In [20]:
lb= LabelEncoder$new()

In [21]:
# c(16:18)
for (j in 16:18) { 
df[,j] = lb$fit_transform(df[,j])
}

In [22]:
#1,5
df[,1] = lb$fit_transform(df[,1])
df[,5] = lb$fit_transform(df[,5])

In [23]:
head(df)

HeartDisease BMI   Smoking AlcoholDrinking Stroke PhysicalHealth MentalHealth
1 0            16.60 Yes     No              0       3             30          
2 0            20.34 No      No              1       0              0          
3 0            26.58 Yes     No              0      20             30          
4 0            24.21 No      No              0       0              0          
5 0            23.71 No      No              0      28              0          
6 1            28.87 Yes     No              0       6              0          
  DiffWalking Sex    AgeCategory Race  Diabetic PhysicalActivity GenHealth
1 No          Female 55-59       White 1        Yes              Very good
2 No          Female 80 or older White 0        Yes              Very good
3 No          Male   65-69       White 1        Yes              Fair     
4 No          Female 75-79       White 0        No               Good     
5 Yes         Female 40-44       White 0        Yes              Very good
6 Yes         Female 75-79       Black 0        No               Fair     
  SleepTime Asthma KidneyDisease SkinCancer
1  5        1      0             1         
2  7        0      0             0         
3  8        1      0             0         
4  6        0      0             1         
5  8        0      0             0         
6 12        0      0             0

## 2.2. GenHealth 라벨 인코딩 (순서O)

In [24]:
table(df[14])


Excellent      Fair      Good      Poor Very good 
    66842     34677     93129     11289    113858 

- STEP1. poor(5) < fair(4) < good(3) < very good(2) < excellent(1)
- 왜냐하면 질병 NO==0, YES==1

In [ ]:
##df[,12]<-ifelse(df[,12] %in% "Yes" , 1,0 )

In [38]:
df[,14]<-ifelse(df[,14]=="Poor",5, 
               ifelse(df[,14]=="Fair",4,
                     ifelse(df[,14]=="Good",3, 
                           ifelse(df[,14]=="Very good",2,1))))

In [39]:
table(df[14]) # 확인


     1      2      3      4      5 
 66842 113858  93129  34677  11289 

# 3. Feature Engineering
- '질병 이력 합' 변수 새로 만들기 

In [40]:
library(dplyr)

Warning message:
"package 'dplyr' was built under R version 3.6.3"
Attaching package: 'dplyr'

The following object is masked from 'package:gridExtra':

    combine

The following objects are masked from 'package:stats':

    filter, lag

The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union



In [42]:
df<- mutate(df, tot= HeartDisease+Stroke+Diabetic+Asthma+KidneyDisease+SkinCancer)
table(df$tot)


     0      1      2      3      4      5      6 
201583  83806  24889   7308   1870    308     31 

# 4. Correlation

In [45]:
#새로운 데이터프레임 생성 
df_n<-df[,c(11,14,19)] 

In [47]:
df_n %>% group_by(Race) %>% summarise(cor=cor( GenHealth,tot ) )

Race                           cor      
1 American Indian/Alaskan Native 0.3672213
2 Asian                          0.2794953
3 Black                          0.3764421
4 Hispanic                       0.3353170
5 Other                          0.3713269
6 White                          0.3652956

In [48]:
df_f<-as.data.frame( df_n %>% group_by(Race) %>% summarise(cor=cor( GenHealth,tot ) )  ) 

In [49]:
#정렬
o<-order(df_f$cor, decreasing=TRUE)
head(df_f[o,])

Race                           cor      
3 Black                          0.3764421
5 Other                          0.3713269
1 American Indian/Alaskan Native 0.3672213
6 White                          0.3652956
4 Hispanic                       0.3353170
2 Asian                          0.2794953

# 5. Conclusion
1. '질병 이력의 합'과 '주관적 건강 인식'의 상관관계는 유의미하다고 할 근거가 부족하다.
2. 두 변수의 상관계수가 가장 큰 인종은 흑인(Black)이며, 값은 0.3764 이다.
3. 두 변수의 상관계수가 가장 작은 인종은 아시안(Asian)이며, 값은 0.2795 이다.
4. 이 차이의 값은 약 0.0969이며 유의미하다고 할 근거가 부족하다. 